lib/services/database/database_service.dart

```diff
@@ -0,0 +1,256 @@
+import 'dart:async';
+
+import "package:collection/collection.dart";
+import 'package:mobileprism/services/database/album_data_entry.dart';
+import 'package:mobileprism/services/database/photo_data_entry.dart';
+import 'package:path/path.dart';
+import 'package:path_provider/path_provider.dart';
+import 'package:sqflite/sqflite.dart';
+
+const String photoDataTableName = "photo_data";
+const String albumDataTableName = "album_data";
+
+late final Database? _database;
+
+class DbNotOpenException implements Exception {}
+
+class FieldsMissingException implements Exception {}
+
+class DuplicateKeyException implements Exception {}
+
+class SqlFilter {
+  final String column;
+  final String operator;
+  final String value;
+  final String? comparator;
+
+  const SqlFilter(this.column, this.operator, this.value, {this.comparator});
+
+  @override
+  String toString() {
+    if (comparator == null) {
+      return '("$column" $operator ?)';
+    } else {
+      return ' $comparator ("$column" $operator ?)';
+    }
+  }
```

Comment:
There is one whitespace at the beginning of line 34 string
```dart
    return comparator == null
        ? '("$column" $operator ?)'
        : ' $comparator ("$column" $operator ?)';
```

Reply:
I know, I added it intentionally.

Reply:
May I ask why?

Reply:
Because the filter would be formatted odd otherwise.